### PR TITLE
VFB-20 Implement the status/action bar for the clients page

### PR DESCRIPTION
**Pull request by Var1377**
This implements the bar that allows the user to add events to the database when parcels are selected.

(no longer a draft)

### CODE REVIEW

File: src/app/clients/ActionBar.tsx [hunk 1]
```diff
@@ -0,0 +1,289 @@
...
+                    type="button"
+                    id="action-button"
+                >
+                    Actions
+                </Button>
+            </OuterDiv>
+        </>
+    );
+};
+
```
Comment (SamYap-Softwire) [inline src/app/clients/ActionBar.tsx:186]: What do you think about splitting this into 2 files here?

File: src/app/clients/ActionBar.tsx [hunk 2]
```diff
@@ -0,0 +1,289 @@
...
+const ModalInner = styled.div`
+    display: flex;
+    flex-direction: column;
+    gap: 1rem;
+    align-items: stretch;
+`;
+
+const StatusText = styled.p`
+    margin-left: 1rem;
+    border-top: 1px solid darkgrey;
```
Comment (SamYap-Softwire) [inline src/app/clients/ActionBar.tsx:213]: Is this going to be a colour from `themes.tsx`?

File: src/app/clients/ActionBar.tsx [hunk 3]
```diff
@@ -0,0 +1,289 @@
...
+                </div>
+                <p>To</p>
+                <div>
+                    {props.data.map((parcel, index) => {
+                        return (
+                            <StatusText key={index}>
+                                {parcel.collection_centre}: (
+                                {parcel.collection_datetime ? (
+                                    new Date(parcel.collection_datetime).toLocaleString()
+                                ) : (
```
Comment (SamYap-Softwire) [inline src/app/clients/ActionBar.tsx:272]: Can we use `&&` here? (only if we don't need the empty fragment)

File: src/app/clients/ActionBar.tsx [hunk 4]
```diff
@@ -0,0 +1,289 @@
...
+                                    <></>
+                                )}
+                                )
+                            </StatusText>
+                        );
+                    })}
+                </div>
+                {props.errorText ? <small>{props.errorText}</small> : <></>}
```
Comment (SamYap-Softwire) [inline src/app/clients/ActionBar.tsx:280]: `&&` here as well please :)

File: src/app/clients/ActionBar.tsx [hunk 5]
```diff
@@ -0,0 +1,289 @@
...
+const StatusText = styled.p`
+    margin-left: 1rem;
+    border-top: 1px solid darkgrey;
+    padding: 1rem 0;
+    &:last-child {
+        border-bottom: 1px solid darkgrey;
+    }
+`;
+
+const SharedModal: React.FC<SharedModalProps> = (props) => {
```
Comment (paopaosoftwire) [inline src/app/clients/ActionBar.tsx:220]: Should the SharedModal and ActionBar be in separate files?

File: src/app/clients/page.tsx [hunk 6]
```diff
@@ -9,7 +8,7 @@ const dataFetch: () => Promise<Schema["clients"][]> = async () => {
 };
 
 const Clients: () => Promise<React.ReactElement> = async () => {
-    const data: Datum[] = await dataFetch();
+    const data = await dataFetch();
```
Comment (paopaosoftwire) [inline src/app/clients/page.tsx:11]: Why was the type removed?

File: src/app/clients/ActionBar.tsx [hunk 7]
```diff
@@ -0,0 +1,289 @@
...
+
+    return (
+        <Modal {...props} header={props.header} headerId={props.headerId}>
+            <ModalInner>
+                <Row>
+                    <p>Date: </p>
+                    <DatePicker
+                        value={date}
+                        defaultValue={date}
+                        onChange={(newDate) =>
```
Comment (paopaosoftwire) [inline src/app/clients/ActionBar.tsx:235]: Is it worth putting these onChange function outside of the component? The function is quite big but its only used once so its up to you:)

**Review by paopaosoftwire — CHANGES_REQUESTED** [review]

File: src/app/clients/ActionBar.cy.tsx [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+
+    const selectedIndices = [0];
+
+    const Component: React.FC<{}> = () => {
+        return (
+            <LocalizationProvider adapterLocale="en-gb" dateAdapter={AdapterDayjs}>
+                <StyleManager>
+                    <ActionBar data={mockData} selected={selectedIndices} />
+                </StyleManager>
+            </LocalizationProvider>
```
Comment (DanWilkins2107) [inline src/app/clients/ActionBar.cy.tsx:29]: Shouldn't the localization provider already be added in themes? or is it different because its a cypress file?

File: src/app/clients/ActionBar.cy.tsx [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+            collection_centre: "Clapham",
+            collection_datetime: null,
+            voucher_number: null,
+            packing_datetime: null,
+        },
+    ];
+
+    const selectedIndices = [0];
+
+    const Component: React.FC<{}> = () => {
```
Comment (DanWilkins2107) [inline src/app/clients/ActionBar.cy.tsx:23]: Could have a better name than component, but also that is me being pedantic lol

File: src/app/clients/ActionBar.tsx [hunk 10]
```diff
@@ -0,0 +1,289 @@
...
+import dayjs, { Dayjs } from "dayjs";
+
+interface Props {
+    selected: number[];
+    data: Schema["parcels"][];
+}
+
+const statuses = [
+    "Request Denied",
+    // probably shouldn't be a status (?)
```
[the remark this answers is not in the document]
Comment (Var1377) [reply src/app/clients/ActionBar.tsx:21]: No Status

File: src/app/clients/ActionBar.tsx [hunk 11]
```diff
@@ -0,0 +1,289 @@
...
+    "No Status",
```
[the remark this answers is not in the document]
Comment (Var1377) [reply src/app/clients/ActionBar.tsx:22]: No clue, I just copied the original site.

File: src/app/clients/page.tsx (hunk 6 above)
Comment (Var1377) [reply src/app/clients/page.tsx:11]: type inference gives it a more specific type and results in no type errors, so having no explicit type would be preferred.

File: src/app/clients/ActionBar.tsx (hunk 7 above)
Comment (DanWilkins2107) [reply src/app/clients/ActionBar.tsx:235]: I would agree with this tbf

**Review by DanWilkins2107 — CHANGES_REQUESTED** [review]

File: src/app/clients/ActionBar.cy.tsx (hunk 9 above)
[the remark this answers is not in the document]
Comment (Var1377) [reply src/app/clients/ActionBar.cy.tsx:23]: We don't but @TomCusack-Softwire says we should keep it there specifically for React.FC

File: src/app/clients/ActionBar.tsx [hunk 12]
```diff
@@ -0,0 +1,289 @@
...
+    gap: 1rem;
+    border-radius: 0.5rem;
+    background-color: ${(props) => props.theme.surfaceBackgroundColor};
+    margin: 1rem;
+`;
+
+const ActionBar: React.FC<Props> = ({ selected, data }) => {
+    const selectedData = Array.from(selected.map((index) => data[index]));
+
+    const [statusAnchorEl, setStatusAnchorEl] = React.useState<null | HTMLElement>(null);
```
[the remark this answers is not in the document]
Comment (Var1377) [reply src/app/clients/ActionBar.tsx:57]: good catch

File: src/app/clients/ActionBar.tsx [hunk 13]
```diff
@@ -0,0 +1,186 @@
...
+                    setModalError(null);
+                }}
+                data={selectedData}
+                status={selectedAction}
+                header="Perform Action"
+                headerId="action-modal-header"
+                onSubmit={submitAction}
+                errorText={modalError}
+            />
+            <Menu
```
Comment (paopaosoftwire) [inline src/app/clients/ActionBar.tsx:120]: Seems like this section and the section below is very similar, would it be better to have a component? or at least have the onClick function separated out.

**Review by paopaosoftwire — COMMENTED** [review]
Looking good!

File: src/app/clients/ActionBar.tsx (hunk 13 above)
Comment (Var1377) [reply src/app/clients/ActionBar.tsx:120]: My instinct would suggest that it would be better to keep it inline although other people's comments/opinions are welcome

Comment (paopaosoftwire) [reply src/app/clients/ActionBar.tsx:120]: I get what you mean. Happy to keep it as it is!

**Review by paopaosoftwire — COMMENTED** [review]
Looks good. Just some questions and suggestions!

File: src/app/clients/ActionBar.tsx [hunk 14]
```diff
@@ -0,0 +1,186 @@
...
+
+        if (error) {
+            setModalError(error.message);
+        } else {
+            setStatusModal(false);
+            setModalError(null);
+        }
+    };
+
+    const submitAction = async (): Promise<void> => {};
```
[the remark this answers is not in the document]
Comment (Var1377) [reply src/app/clients/ActionBar.tsx:90]: For whenever we implement submitting actions. I'll add a TODO

File: src/app/clients/ActionBar.tsx [hunk 15]
```diff
@@ -0,0 +1,188 @@
...
+
+interface Props {
+    selected: number[];
+    data: Schema["parcels"][];
+}
+
+const statuses = [
+    "Request Denied",
+    // probably shouldn't be a status (?)
+    "No Status",
```
Comment (DanWilkins2107) [inline src/app/clients/ActionBar.tsx:21]: Idk if this would make more sense as `"Reset Status"` if that is what is meant to be? If `No Status` is what it's meant to be that's fine though

File: src/app/clients/ActionBar.tsx [hunk 16]
```diff
@@ -0,0 +1,188 @@
...
+    "Delete Request",
+];
+
+const OuterDiv = styled.div`
+    display: flex;
+    padding: 1rem;
+    gap: 1rem;
+    border-radius: 0.5rem;
+    background-color: ${(props) => props.theme.surfaceBackgroundColor};
+    margin: 1rem;
```
Comment (DanWilkins2107) [inline src/app/clients/ActionBar.tsx:50]: Why use both padding and margin? Just cos surely `padding: 2rem` would have the same effect. Or is this just conventional to use both?

File: src/app/clients/ActionBar.tsx [hunk 17]
```diff
@@ -0,0 +1,188 @@
...
+`;
+
+const ActionBar: React.FC<Props> = ({ selected, data }) => {
+    const selectedData = Array.from(selected.map((index) => data[index]));
+
+    const [statusAnchorEl, setStatusAnchorEl] = useState<null | HTMLElement>(null);
```
Comment (DanWilkins2107) [inline src/app/clients/ActionBar.tsx:56]: Is it necessary to shorten this to El? Like I'm guessing it means Element but I don't see why you wouldn't just have it as `statusAnchorElement` for clarity.

File: src/app/clients/ActionBarModal.tsx [hunk 18]
```diff
@@ -0,0 +1,106 @@
...
+    gap: 1rem;
+    align-items: stretch;
+`;
+
+const StatusText = styled.p`
+    margin-left: 1rem;
+    border-top: 1px solid darkgrey;
+    padding: 1rem 0;
+    &:last-child {
+        border-bottom: 1px solid darkgrey;
```
Comment (DanWilkins2107) [inline src/app/clients/ActionBarModal.tsx:38]: Can these be colours from the theme?

**Review by DanWilkins2107 — CHANGES_REQUESTED** [review]
Some small stylistic changes to make, but fine other than that.

File: src/app/clients/ActionBar.tsx (hunk 15 above)
Comment (Var1377) [reply src/app/clients/ActionBar.tsx:21]: "No Status" was copied from their current implementation, the comment was an expression of my own confusion

File: src/app/clients/ActionBar.tsx (hunk 16 above)
Comment (Var1377) [reply src/app/clients/ActionBar.tsx:50]: They aren't the same with a `background-color` or `background-image`, the decoration is applied to the padding area but not the margin area.

File: src/app/clients/ActionBar.tsx [hunk 19]
```diff
@@ -0,0 +1,188 @@
...
+
+    const [statusModal, setStatusModal] = useState(false);
+    const [actionModal, setActionModal] = useState(false);
+
+    const [modalError, setModalError] = useState<string | null>(null);
+
+    const submitStatus = async (date: Dayjs): Promise<void> => {
+        const toInsert = selectedData
+            .map((parcel) => {
+                const event_name = selectedStatus!;
```
[the remark this answers is not in the document]
Comment (Var1377) [reply src/app/clients/ActionBar.tsx:69]: muscle memory...

nice catch

[the remark this answers is not in the document]
Comment (Var1377) [reply src/app/clients/ActionBar.tsx:69]: Nevermind, it's because those are the same names of the database columns

File: src/app/clients/ActionBarModal.tsx (hunk 18 above)
Comment (Var1377) [reply src/app/clients/ActionBarModal.tsx:38]: I'll add a TODO for VFB-16

File: src/app/clients/ActionBar.tsx [hunk 20]
```diff
@@ -142,9 +142,9 @@ const ActionBar: React.FC<Props> = ({ selected, data }) => {
                 </MenuList>
             </Menu>
             <Menu
-                open={actionAnchorEl !== null}
-                onClose={() => setActionAnchorEl(null)}
```
Comment (DanWilkins2107) [inline src/app/clients/ActionBar.tsx]: Try to fix the navBar taking up the whole screen on open and close :)

**Review by DanWilkins2107 — CHANGES_REQUESTED** [review]

File: src/app/clients/ActionBar.tsx (hunk 20 above)
Comment (Var1377) [reply src/app/clients/ActionBar.tsx]: I can't reproduce this on any browser under any viewport size...

**Review by paopaosoftwire — APPROVED** [review]
This looks fine now! Great stuff. I see some comments for VFB16 but thats already merged so I think that should be removed if you rebase main. And I think Tom doesn't like TODOs with a colon 😂. But I'll approve it :)

File: src/app/clients/ActionBar.cy.tsx [hunk 21]
```diff
@@ -0,0 +1,84 @@
+import ActionBar from "@/app/clients/ActionBar";
+import React from "react";
+import StyleManager from "@/app/themes";
+import { ClientsTableRow } from "./getClientsTableData";
```
Comment (SamYap-Softwire) [inline src/app/clients/ActionBar.cy.tsx:4]: @ import please :)

File: src/app/clients/ActionBar.tsx (hunk 11 above)
[the remark this answers is not in the document]
Comment (SamYap-Softwire) [reply src/app/clients/ActionBar.tsx:23]: might be worth looking into it?

File: src/app/clients/ActionBar.tsx (hunk 10 above)
[the remark this answers is not in the document]
Comment (SamYap-Softwire) [reply src/app/clients/ActionBar.tsx:22]: Yea I don't know if it should be a status 🤷 (not a very helpful comment ...)

File: src/app/clients/ActionBar.tsx [hunk 22]
```diff
@@ -0,0 +1,193 @@
...
+        if (error) {
+            setModalError(error.message);
+        } else {
+            setStatusModal(false);
+            setModalError(null);
+        }
+    };
+
+    const submitAction = async (): Promise<void> => {
+        // TODO: implement whatever actions are meant to do
```
Comment (SamYap-Softwire) [inline src/app/clients/ActionBar.tsx:96]: Is this a note to self or for a future ticket?

File: src/app/clients/ActionBarModal.tsx [hunk 23]
```diff
@@ -0,0 +1,109 @@
...
+`;
+
+const ModalInner = styled.div`
+    display: flex;
+    flex-direction: column;
+    gap: 1rem;
+    align-items: stretch;
+`;
+
+// TODO VFB-16 Make these border colours change on dark and light themes
```
Comment (SamYap-Softwire) [inline src/app/clients/ActionBarModal.tsx:33]: Now that themes has been merged into main, can you rebase and update this?

File: src/app/clients/ActionBarModal.tsx [hunk 24]
```diff
@@ -0,0 +1,109 @@
...
+                <Row>
+                    <p>Time: </p>
+                    <TimePicker value={date} onChange={onTimeChange} disableFuture />
+                </Row>
+                <p>Applying:</p>
+                <div>
+                    <StatusText key={props.status}>{props.status}</StatusText>
+                </div>
+                <p>To</p>
+                <div>
```
Comment (SamYap-Softwire) [inline src/app/clients/ActionBarModal.tsx:86]: what's the difference between a `<div></div>` and a `<></>`?

File: src/app/clients/ActionBarModal.tsx [hunk 25]
```diff
@@ -0,0 +1,109 @@
...
+            date
+                .set("hour", newDate?.hour() ?? date.hour())
+                .set("minute", newDate?.minute() ?? date.minute())
+        );
+
+    return (
+        <Modal {...props} header={props.header} headerId={props.headerId}>
+            <ModalInner>
+                <Row>
+                    <p>Date: </p>
```
Comment (SamYap-Softwire) [inline src/app/clients/ActionBarModal.tsx:69]: Can this work without a `<p>` tag? (is it `<p>` by default?)

File: src/app/clients/ClientsPage.tsx [hunk 26]
```diff
@@ -19,6 +19,7 @@ import { Schema } from "@/supabase";
 import TableSurface from "@/components/Tables/TableSurface";
 import { CenterComponent } from "@/components/Form/formStyling";
 import Button from "@mui/material/Button";
+import ActionBar from "./ActionBar";
```
Comment (SamYap-Softwire) [inline src/app/clients/ClientsPage.tsx:22]: @ import

File: src/components/Tables/Table.tsx [hunk 27]
```diff
@@ -151,10 +153,19 @@ const Table: React.FC<Props> = ({
 
     const [filterText, setFilterText] = useState<FilterText>({});
 
-    const [selectCheckBoxes, setSelectCheckBoxes] = useState(
+    const [selectCheckBoxes, _setSelectCheckBoxes] = useState(
         new Array<boolean>(data.length).fill(false)
     );
 
+    const setSelectCheckBoxes = (newSelection: boolean[]): void => {
```
Comment (SamYap-Softwire) [inline src/components/Tables/Table.tsx:160]: (continuing from the previous comment...) and rename this?

File: src/components/Tables/Table.tsx [hunk 28]
```diff
@@ -151,10 +153,19 @@ const Table: React.FC<Props> = ({
 
     const [filterText, setFilterText] = useState<FilterText>({});
 
-    const [selectCheckBoxes, setSelectCheckBoxes] = useState(
+    const [selectCheckBoxes, _setSelectCheckBoxes] = useState(
```
Comment (SamYap-Softwire) [inline src/components/Tables/Table.tsx:156]: can we keep this as `setSelectCheckBoxes` because I think we've set underscore at the start of variable names to mean unused variable.

**Review by SamYap-Softwire — CHANGES_REQUESTED** [review]
Minor changes

File: src/app/clients/ActionBarModal.tsx (hunk 24 above)
Comment (Var1377) [reply src/app/clients/ActionBarModal.tsx:86]: By having it as an element, it affects the flex layout (specifically where the flex gap appears)

File: src/components/Tables/Table.tsx (hunk 28 above)
Comment (DanWilkins2107) [reply src/components/Tables/Table.tsx:156]: Also generally why would you rename it (like is there a reason why this is stylistically better)

File: src/components/Tables/Table.tsx [hunk 29]
```diff
@@ -164,7 +175,8 @@ const Table: React.FC<Props> = ({
     };
 
     const toggleAllCheckBox = (): void => {
-        setSelectCheckBoxes(new Array<boolean>(data.length).fill(!selectAllCheckBox));
+        const newSelection = new Array<boolean>(data.length).fill(!selectAllCheckBox);
+        setSelectCheckBoxes(newSelection);
```
Comment (DanWilkins2107) [inline src/components/Tables/Table.tsx:179]: Is this just a new function? Or is it the same as the one with the underscore at the front?

**Review by DanWilkins2107 — CHANGES_REQUESTED** [review]

File: src/app/clients/ActionBarModal.tsx (hunk 25 above)
Comment (Var1377) [reply src/app/clients/ActionBarModal.tsx]: It can, they've been removed although I'm not sure if it's now better or not

Comment (SamYap-Softwire) [reply src/app/clients/ActionBarModal.tsx]: i prefer it this way (removed) so thanks :)

**Review by SamYap-Softwire — APPROVED** [review]
looks good!

**Review by DanWilkins2107 — APPROVED** [review]

File: src/app/clients/ActionBar.cy.tsx [hunk 30]
```diff
@@ -0,0 +1,119 @@
...
+
+            it("should have a modal with date and time pickers", () => {
+                cy.get(`#${type}-button`).click();
+                cy.get(`#${type}-menu`).should("exist");
+                cy.get(`#${type}-menu`).contains(example).click();
+                cy.get(`#${type}-modal-header`).should("exist");
+                cy.get(".MuiPaper-root").contains("Date");
+                cy.get(".MuiPaper-root").contains("Time");
+                const dateString = new Date().toLocaleDateString("en-GB");
+                cy.get("input[value='" + dateString + "']").should("exist");
```
Comment (TomCusack-Softwire) [inline src/app/clients/ActionBar.cy.tsx:90]: can these all be template strings - `` `input[value='${dateString}']` ``

File: src/app/clients/ActionBarModal.tsx [hunk 31]
```diff
@@ -0,0 +1,106 @@
...
+    margin-left: 1rem;
+    border-top: 1px solid darkgrey;
+    padding: 1rem 0;
+    &:last-child {
+        border-bottom: 1px solid darkgrey;
+    }
+`;
+
+const ActionBarModal: React.FC<SharedModalProps> = (props) => {
+    const [date, setDate] = useState(dayjs(Date()));
```
Comment (TomCusack-Softwire) [inline src/app/clients/ActionBarModal.tsx:43]: do you want `new Date()` when calling each of these Date constructors?

File: src/app/clients/ClientsPage.tsx [hunk 32]
```diff
@@ -166,6 +168,7 @@ const ClientsPage: React.FC<Props> = (props) => {
                     columnStyleOptions={clientTableColumnStyleOptions}
                     onRowClick={onClientTableRowClick}
                     checkboxes={true}
+                    onRowSelection={setSelected}
```
Comment (TomCusack-Softwire) [inline src/app/clients/ClientsPage.tsx:171]: how does this work with pagination? can you select rows 1 and 20 if only 10 are shown at once?

File: src/app/themes.tsx [hunk 33]
```diff
@@ -185,11 +188,13 @@ const StyleManager: React.FC<Props> = ({ children, theme = lightTheme }) => {
         );
 
     return (
-        <ThemeUpdateContext.Provider value={handleThemeChange}>
-            <ThemeProvider theme={chosenTheme}>
-                <MaterialAndGlobalStyle>{themedChildren}</MaterialAndGlobalStyle>
-            </ThemeProvider>
-        </ThemeUpdateContext.Provider>
+        <LocalizationProvider dateAdapter={AdapterDayjs} adapterLocale="en-gb">
```
Comment (TomCusack-Softwire) [inline src/app/themes.tsx:191]: don't think this should be in `themes.tsx` but instead `layout.tsx`

**Review by TomCusack-Softwire — APPROVED** [review]

File: src/app/clients/ClientsPage.tsx (hunk 32 above)
Comment (Var1377) [reply src/app/clients/ClientsPage.tsx:171]: With the current implementation of pagination it works fine (it uses rowIdx which is always unique even between pages) but this might break once @paopaosoftwire changes how pagination works
